File: app/src/test/java/za/co/riggaroo/datecountdown/repository/EventRepositoryImplTest.java
package za.co.riggaroo.datecountdown.repository;

import android.arch.core.executor.testing.InstantTaskExecutorRule;
import android.arch.lifecycle.MutableLiveData;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.List;

import za.co.riggaroo.datecountdown.FakeEventDataGenerator;
import za.co.riggaroo.datecountdown.LiveDataUtils;
import za.co.riggaroo.datecountdown.data.dao.EventDao;
import za.co.riggaroo.datecountdown.data.entity.Event;

import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class EventRepositoryImplTest {

    @Mock
    private EventDao eventDao;
    @Rule
    public InstantTaskExecutorRule instantExecutorRule = new InstantTaskExecutorRule();

    private EventRepository eventRepository;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        eventRepository = new EventRepositoryImpl(eventDao);
    }

    @Test
    public void addEvent_TriggersDbAdd(){
        Event event = FakeEventDataGenerator.getFakeEvent();
        eventRepository.addEvent(event).test().onComplete();

        verify(eventDao).addEvent(event);
    }
    @Test
    public void addEvent_EventNull_ThrowsIllegalArgumentException(){
        eventRepository.addEvent(null)
                .test()
                .assertError(IllegalArgumentException.class);

        verify(eventDao, never()).addEvent(null);
    }

    @Test
    public void getEvent_DaoGetEvents() throws InterruptedException {

        MutableLiveData<List<Event>> fakeEvents = FakeEventDataGenerator.getEventListMutableData();
        when(eventDao.getEvents(any())).thenReturn(fakeEvents);

        List<Event> eventsReturned = LiveDataUtils.getValue(eventRepository.getEvents());

        verify(eventDao).getEvents(any());
        assertArrayEquals(fakeEvents.getValue().toArray(), eventsReturned.toArray());
    }

    @Test
    public void deleteEvent_TriggersDbDelete(){
        Event event = FakeEventDataGenerator.getFakeEvent();
        eventRepository.deleteEvent(event).test().onComplete();

        verify(eventDao).deleteEvent(event);
    }
    @Test
    public void deleteEvent_EventNull_ThrowsIllegalArgumentException(){
        eventRepository.deleteEvent(null)
                .test()
                .assertError(IllegalArgumentException.class);

        verify(eventDao, never()).addEvent(null);
    }

}

File: app/src/test/java/za/co/riggaroo/datecountdown/repository/EventRepositoryImplTest.kt
package za.co.riggaroo.datecountdown.repository

import android.arch.core.executor.testing.InstantTaskExecutorRule
import org.junit.Assert.assertArrayEquals
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import za.co.riggaroo.datecountdown.FakeEventDataGenerator
import za.co.riggaroo.datecountdown.LiveDataUtils
import za.co.riggaroo.datecountdown.data.dao.EventDao

class EventRepositoryImplTest {

    @Mock
    private lateinit var eventDao: EventDao

    @Rule
    @JvmField
    var instantExecutorRule = InstantTaskExecutorRule()

    private lateinit var eventRepository: EventRepository

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        eventRepository = EventRepositoryImpl(eventDao)
    }

    @Test
    fun addEvent_TriggersDbAdd() {
        val event = FakeEventDataGenerator.getFakeEvent()
        eventRepository.addEvent(event).test().onComplete()

        verify<EventDao>(eventDao).addEvent(event)
    }

    @Test
    fun addEvent_EventNull_ThrowsIllegalArgumentException() {
        eventRepository.addEvent(null)
                .test()
                .assertError(IllegalArgumentException::class.java)

        verify<EventDao>(eventDao, never()).addEvent(any())
    }

    @Test
    @Throws(InterruptedException::class)
    fun getEvent_DaoGetEvents() {

        val fakeEvents = FakeEventDataGenerator.getEventListMutableData()
        `when`(eventDao.getEvents(any())).thenReturn(fakeEvents)

        val eventsReturned = LiveDataUtils.getValue(eventRepository.getEvents())

        verify(eventDao).getEvents(any())
        assertArrayEquals(fakeEvents.value?.toTypedArray(), eventsReturned.toTypedArray())
    }

    @Test
    fun deleteEvent_TriggersDbDelete() {
        val event = FakeEventDataGenerator.getFakeEvent()
        eventRepository.deleteEvent(event).test().onComplete()

        verify<EventDao>(eventDao).deleteEvent(event)
    }

    @Test
    fun deleteEvent_EventNull_ThrowsIllegalArgumentException() {
        eventRepository.deleteEvent(null)
                .test()
                .assertError(IllegalArgumentException::class.java)
    }


    private fun <T> any(): T {
        Mockito.any<T>()
        return uninitialized()
    }

    private fun <T> uninitialized(): T = null as T
}